compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirFunctionParameterChecker.kt

```diff
@@ -20,6 +24,25 @@ import org.jetbrains.kotlin.types.AbstractTypeChecker
 object FirFunctionParameterChecker : FirFunctionChecker() {
     override fun check(declaration: FirFunction<*>, context: CheckerContext, reporter: DiagnosticReporter) {
         checkVarargParameters(declaration, context, reporter)
+        checkParameterTypes(declaration, context, reporter)
+    }
+
+    private fun checkParameterTypes(declaration: FirFunction<*>, context: CheckerContext, reporter: DiagnosticReporter) {
+        for (valueParameter in declaration.valueParameters) {
+            val returnTypeRef = valueParameter.returnTypeRef
+            if (returnTypeRef !is FirErrorTypeRef) continue
+            // type problems on real source are already reported by ConeDiagnostic.toFirDiagnostics
+            if (returnTypeRef.source?.kind !is FirFakeSourceElementKind?) continue
```

Comment:
The `?` is intentional, it's basically a null check plus a type check.

To your second point, yep you are right. For this particular check I can remove this if condition. Originally I put it there since I was thinking maybe we can have some other checks here in future. But I agree it's confusing in the current code as it is. So let me remove it.

Reply:
Actually, I think it's still a good check to have just to make sure we don't report duplicated error when this code is changed in future. But let me make the condition more readable.

Reply:
`returnTypeRef.source?.kind == FirRealSourceElementKind` is much more readable, thank you!